File: src/scoracle_data/external/twitter.py
"""Twitter/X API v2 client for fetching tweets about sports entities."""

import os
import logging
from dataclasses import dataclass
from datetime import datetime
from typing import Any

from .base import BaseExternalClient, ExternalAPIError

logger = logging.getLogger(__name__)


@dataclass
class TweetAuthor:
    """Tweet author information."""
    id: str
    username: str
    name: str
    verified: bool
    profile_image_url: str | None


@dataclass
class TweetMetrics:
    """Tweet engagement metrics."""
    likes: int
    retweets: int
    replies: int
    quotes: int


@dataclass
class Tweet:
    """Normalized tweet data."""
    id: str
    text: str
    author: TweetAuthor
    created_at: str
    metrics: TweetMetrics
    url: str


class TwitterClient(BaseExternalClient):
    """
    Twitter/X API v2 client.

    Uses the recent search endpoint to find tweets about players/teams.
    Rate limit: 450 requests per 15 minutes (Essential tier).
    """

    def __init__(self, bearer_token: str | None = None):
        """
        Initialize Twitter client.

        Args:
            bearer_token: Twitter API v2 bearer token. If None, reads from
                          TWITTER_BEARER_TOKEN environment variable.
        """
        super().__init__(
            base_url="https://api.twitter.com/2",
            rate_limit=(450, 900),  # 450 requests per 15 minutes
            timeout=15.0,
        )
        self.bearer_token = bearer_token or os.getenv("TWITTER_BEARER_TOKEN", "")

    def _get_auth_headers(self) -> dict[str, str]:
        """Return Bearer token auth header."""
        return {"Authorization": f"Bearer {self.bearer_token}"}

    def is_configured(self) -> bool:
        """Check if bearer token is configured."""
        return bool(self.bearer_token)

    def _build_search_query(self, query: str, sport: str | None = None) -> str:
        """
        Build Twitter search query with sport context.

        Args:
            query: Base search query (player/team name)
            sport: Optional sport for context (NBA, NFL, FOOTBALL)

        Returns:
            Enhanced search query
        """
        # Clean the query
        clean_query = query.strip()

        # Add sport-specific context
        sport_context = {
            "NBA": "basketball OR NBA",
            "NFL": "football OR NFL",
            "FOOTBALL": "soccer OR football OR Premier League OR Champions League",
        }

        if sport and sport.upper() in sport_context:
            return f'"{clean_query}" ({sport_context[sport.upper()]}) -is:retweet lang:en'

        return f'"{clean_query}" -is:retweet lang:en'

    def _parse_tweet(self, tweet_data: dict, users_map: dict[str, dict]) -> Tweet:
        """Parse raw API response into Tweet dataclass."""
        author_id = tweet_data.get("author_id", "")
        author_data = users_map.get(author_id, {})

        author = TweetAuthor(
            id=author_id,
            username=author_data.get("username", "unknown"),
            name=author_data.get("name", "Unknown"),
            verified=author_data.get("verified", False),
            profile_image_url=author_data.get("profile_image_url"),
        )

        public_metrics = tweet_data.get("public_metrics", {})
        metrics = TweetMetrics(
            likes=public_metrics.get("like_count", 0),
            retweets=public_metrics.get("retweet_count", 0),
            replies=public_metrics.get("reply_count", 0),
            quotes=public_metrics.get("quote_count", 0),
        )

        tweet_id = tweet_data.get("id", "")
        return Tweet(
            id=tweet_id,
            text=tweet_data.get("text", ""),
            author=author,
            created_at=tweet_data.get("created_at", ""),
            metrics=metrics,
            url=f"https://twitter.com/{author.username}/status/{tweet_id}",
        )

    async def search(
        self,
        query: str,
        sport: str | None = None,
        limit: int = 10,
    ) -> dict[str, Any]:
        """
        Search for recent tweets about a sports entity.

        Args:
            query: Search query (player name, team name)
            sport: Optional sport context (NBA, NFL, FOOTBALL)
            limit: Maximum number of results (1-100)

        Returns:
            Dictionary with tweets and metadata
        """
        if not self.is_configured():
            raise ExternalAPIError(
                "Twitter API not configured. Set TWITTER_BEARER_TOKEN environment variable.",
                code="SERVICE_UNAVAILABLE",
                status_code=503,
            )

        search_query = self._build_search_query(query, sport)
        limit = min(max(1, limit), 100)  # Clamp to 1-100

        params = {
            "query": search_query,
            "max_results": limit,
            "tweet.fields": "created_at,public_metrics,author_id",
            "user.fields": "username,name,verified,profile_image_url",
            "expansions": "author_id",
        }

        try:
            response = await self.get("/tweets/search/recent", params=params)
        except ExternalAPIError:
            raise
        except Exception as e:
            logger.error(f"Twitter search failed: {e}")
            raise ExternalAPIError(f"Twitter search failed: {str(e)}")

        # Build users map for author lookup
        users_map = {}
        includes = response.get("includes", {})
        for user in includes.get("users", []):
            users_map[user["id"]] = user

        # Parse tweets
        tweets = []
        for tweet_data in response.get("data", []):
            try:
                tweet = self._parse_tweet(tweet_data, users_map)
                tweets.append({
                    "id": tweet.id,
                    "text": tweet.text,
                    "author": {
                        "username": tweet.author.username,
                        "name": tweet.author.name,
                        "verified": tweet.author.verified,
                        "profile_image_url": tweet.author.profile_image_url,
                    },
                    "created_at": tweet.created_at,
                    "metrics": {
                        "likes": tweet.metrics.likes,
                        "retweets": tweet.metrics.retweets,
                        "replies": tweet.metrics.replies,
                    },
                    "url": tweet.url,
                })
            except Exception as e:
                logger.warning(f"Failed to parse tweet: {e}")
                continue

        meta = response.get("meta", {})

        return {
            "query": query,
            "sport": sport,
            "tweets": tweets,
            "meta": {
                "result_count": meta.get("result_count", len(tweets)),
                "newest_id": meta.get("newest_id"),
                "oldest_id": meta.get("oldest_id"),
            },
        }

File: internal/external/twitter.go
package external

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const (
	twitterBaseURL    = "https://api.twitter.com/2"
	twitterTimeout    = 15 * time.Second
	twitterCacheTTL   = 1 * time.Hour // default feed cache TTL
	twitterMaxResults = 100
)

// ---------------------------------------------------------------------------
// Tweet types — normalized output
// ---------------------------------------------------------------------------

// TweetAuthor is the author of a tweet.
type TweetAuthor struct {
	Username        string  `json:"username"`
	Name            string  `json:"name"`
	Verified        bool    `json:"verified"`
	ProfileImageURL *string `json:"profile_image_url"`
}

// TweetMetrics holds engagement metrics.
type TweetMetrics struct {
	Likes    int `json:"likes"`
	Retweets int `json:"retweets"`
	Replies  int `json:"replies"`
}

// Tweet is a normalized tweet.
type Tweet struct {
	ID        string       `json:"id"`
	Text      string       `json:"text"`
	Author    TweetAuthor  `json:"author"`
	CreatedAt string       `json:"created_at"`
	Metrics   TweetMetrics `json:"metrics"`
	URL       string       `json:"url"`
}

// ---------------------------------------------------------------------------
// TwitterService — list tweets client with in-memory feed caching
// ---------------------------------------------------------------------------

// TwitterService fetches and caches tweets from a curated X List.
type TwitterService struct {
	bearerToken string
	listID      string
	cacheTTL    time.Duration
	httpClient  *http.Client

	mu             sync.RWMutex
	cachedTweets   []Tweet
	cacheTimestamp time.Time
}

// NewTwitterService creates a twitter service. bearerToken and listID may be empty.
func NewTwitterService(bearerToken, listID string) *TwitterService {
	return &TwitterService{
		bearerToken: bearerToken,
		listID:      listID,
		cacheTTL:    twitterCacheTTL,
		httpClient: &http.Client{
			Timeout: twitterTimeout,
		},
	}
}

// IsConfigured reports whether bearer token and list ID are both set.
func (s *TwitterService) IsConfigured() bool {
	return s.bearerToken != "" && s.listID != ""
}

// CacheTTLSeconds returns the cache TTL in seconds.
func (s *TwitterService) CacheTTLSeconds() int {
	return int(s.cacheTTL.Seconds())
}

// Status returns service configuration status.
func (s *TwitterService) Status() map[string]interface{} {
	return map[string]interface{}{
		"service":                    "twitter",
		"configured":                 s.bearerToken != "",
		"journalist_list_configured": s.listID != "",
		"journalist_list_id":         s.listID,
		"feed_cache_ttl_seconds":     s.CacheTTLSeconds(),
		"rate_limit":                 "900 requests / 15 min (List endpoint)",
		"note":                       "Only journalist-feed endpoint available. Generic search removed to ensure content quality.",
	}
}

// GetJournalistFeed searches the cached journalist feed for query mentions.
func (s *TwitterService) GetJournalistFeed(query, sport string, limit int) (map[string]interface{}, error) {
	if s.bearerToken == "" {
		return nil, fmt.Errorf("Twitter API not configured. Set TWITTER_BEARER_TOKEN.")
	}
	if s.listID == "" {
		return nil, fmt.Errorf("Twitter journalist list not configured. Set TWITTER_JOURNALIST_LIST_ID.")
	}
	if limit < 1 {
		limit = 10
	}
	if limit > 50 {
		limit = 50
	}

	allTweets, feedFromCache, err := s.getOrRefreshFeed()
	if err != nil {
		return nil, err
	}

	// Filter for query matches (case-insensitive substring).
	queryLower := strings.ToLower(query)
	var filtered []Tweet
	for _, t := range allTweets {
		if strings.Contains(strings.ToLower(t.Text), queryLower) {
			filtered = append(filtered, t)
		}
	}

	if len(filtered) > limit {
		filtered = filtered[:limit]
	}

	var sportVal interface{} = sport
	if sport == "" {
		sportVal = nil
	}

	return map[string]interface{}{
		"query":  query,
		"sport":  sportVal,
		"tweets": filtered,
		"meta": map[string]interface{}{
			"result_count":      len(filtered),
			"feed_cached":       feedFromCache,
			"feed_size":         len(allTweets),
			"cache_ttl_seconds": s.CacheTTLSeconds(),
		},
	}, nil
}

// ---------------------------------------------------------------------------
// Internal — feed fetch + cache
// ---------------------------------------------------------------------------

// getOrRefreshFeed returns the cached feed or fetches a fresh one.
// Returns (tweets, fromCache, error).
func (s *TwitterService) getOrRefreshFeed() ([]Tweet, bool, error) {
	s.mu.RLock()
	if s.cachedTweets != nil && time.Since(s.cacheTimestamp) < s.cacheTTL {
		tweets := s.cachedTweets
		s.mu.RUnlock()
		return tweets, true, nil
	}
	s.mu.RUnlock()

	// Fetch fresh feed.
	tweets, err := s.fetchListTweets()
	if err != nil {
		return nil, false, err
	}

	s.mu.Lock()
	s.cachedTweets = tweets
	s.cacheTimestamp = time.Now()
	s.mu.Unlock()

	return tweets, false, nil
}

// fetchListTweets calls GET /2/lists/{id}/tweets.
func (s *TwitterService) fetchListTweets() ([]Tweet, error) {
	params := url.Values{}
	params.Set("max_results", fmt.Sprintf("%d", twitterMaxResults))
	params.Set("tweet.fields", "created_at,public_metrics,author_id")
	params.Set("user.fields", "username,name,verified,profile_image_url")
	params.Set("expansions", "author_id")

	u := fmt.Sprintf("%s/lists/%s/tweets?%s", twitterBaseURL, s.listID, params.Encode())

	req, err := http.NewRequest("GET", u, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+s.bearerToken)

	resp, err := s.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("twitter API error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("twitter API HTTP %d", resp.StatusCode)
	}

	var apiResp twitterAPIResponse
	if err := json.NewDecoder(resp.Body).Decode(&apiResp); err != nil {
		return nil, fmt.Errorf("twitter decode error: %w", err)
	}

	return formatTweets(&apiResp), nil
}

// ---------------------------------------------------------------------------
// Twitter API response types
// ---------------------------------------------------------------------------

type twitterAPIResponse struct {
	Data     []twitterTweetData `json:"data"`
	Includes struct {
		Users []twitterUserData `json:"users"`
	} `json:"includes"`
	Meta struct {
		ResultCount int    `json:"result_count"`
		NewestID    string `json:"newest_id"`
		OldestID    string `json:"oldest_id"`
	} `json:"meta"`
}

type twitterTweetData struct {
	ID            string `json:"id"`
	Text          string `json:"text"`
	AuthorID      string `json:"author_id"`
	CreatedAt     string `json:"created_at"`
	PublicMetrics struct {
		LikeCount    int `json:"like_count"`
		RetweetCount int `json:"retweet_count"`
		ReplyCount   int `json:"reply_count"`
		QuoteCount   int `json:"quote_count"`
	} `json:"public_metrics"`
}

type twitterUserData struct {
	ID              string  `json:"id"`
	Username        string  `json:"username"`
	Name            string  `json:"name"`
	Verified        bool    `json:"verified"`
	ProfileImageURL *string `json:"profile_image_url"`
}

// formatTweets converts the raw API response into normalized Tweet objects.
func formatTweets(resp *twitterAPIResponse) []Tweet {
	usersMap := make(map[string]*twitterUserData, len(resp.Includes.Users))
	for i := range resp.Includes.Users {
		usersMap[resp.Includes.Users[i].ID] = &resp.Includes.Users[i]
	}

	tweets := make([]Tweet, 0, len(resp.Data))
	for _, td := range resp.Data {
		user := usersMap[td.AuthorID]
		author := TweetAuthor{
			Username: "unknown",
			Name:     "Unknown",
		}
		if user != nil {
			author = TweetAuthor{
				Username:        user.Username,
				Name:            user.Name,
				Verified:        user.Verified,
				ProfileImageURL: user.ProfileImageURL,
			}
		}

		tweets = append(tweets, Tweet{
			ID:        td.ID,
			Text:      td.Text,
			Author:    author,
			CreatedAt: td.CreatedAt,
			Metrics: TweetMetrics{
				Likes:    td.PublicMetrics.LikeCount,
				Retweets: td.PublicMetrics.RetweetCount,
				Replies:  td.PublicMetrics.ReplyCount,
			},
			URL: fmt.Sprintf("https://twitter.com/%s/status/%s", author.Username, td.ID),
		})
	}

	return tweets
}
